chalice/deploy/packager.py

```diff
@@ -208,3 +205,258 @@ def _add_chalice_lib_if_needed(self, project_dir, zip_fileobj):
                         self._CHALICE_LIB_DIR,
                         fullpath[len(libdir) + 1:])
                     zip_fileobj.write(fullpath, zip_path)
+
+
+class DependencyBuilder(object):
+    """Build site-packages by manually downloading and unpacking whls.
+
+    Pip is used to download all the dependency sdists. Then wheels that
+    compatible with lambda are downloaded. Any source packages that do not
+    have a matching wheel file are built into a wheel and that file is checked
+    for compatibility with the lambda python runtime environment.
+
+    All compatible wheels that are downloaded/built this way are unpacked
+    into a site-packages directory, to be included in the bundle by the
+    packager.
+    """
+    _MANYLINUX_COMPATIBLE_PLAT = {'any', 'linux_x86_64', 'manylinux1_x86_64'}
+
+    def __init__(self, osutils, pip_runner=None):
+        self._osutils = osutils
+        if pip_runner is None:
+            pip_runner = PipRunner(pip)
+        self._pip = pip_runner
+
+    def _valid_lambda_whl(self, filename):
+        # type: (str) -> bool
+        whl = filename[:-4]
+        implementation, abi, platform = whl.split('-')[-3:]
+        # Verify platform is compatible
+        if platform not in self._MANYLINUX_COMPATIBLE_PLAT:
+            return False
+        # Verify that the ABI is compatible with lambda. Either none or the
+        # correct type for the python version cp27mu for py27 and cp36m for
+        # py36.
+        if abi == 'none':
+            return True
+        prefix_version = implementation[:3]
+        if prefix_version == 'cp3':
+            # Deploying python 3 function which means we need cp36m abi
+            return abi == 'cp36m'
+        elif prefix_version == 'cp2':
+            # Deploying to python 2 function which means we need cp27mu abi
+            return abi == 'cp27mu'
+        # Don't know what we have but it didn't pass compatibility tests.
+        return False
+
+    def _has_at_least_one_package(self, filename):
+        # type: (str) -> bool
+        if not os.path.isfile(filename):
+            return False
+        with open(filename, 'r') as f:
+            # This is meant to be a best effort attempt.
+            # This can return True and still have no packages
+            # actually being specified, but those aren't common
+            # cases.
+            for line in f:
+                line = line.strip()
+                if line and not line.startswith('#'):
+                    return True
+        return False
+
+    @contextlib.contextmanager
+    def _tempdir(self):
+        # type: () -> Any
+        tempdir = tempfile.mkdtemp()
+        try:
+            yield tempdir
+        finally:
+            shutil.rmtree(tempdir)
+
+    def _download_dependencies(self, directory, requirements_file):
```

Comment:
It might be worth breaking each of these individual steps into separate internal methods so the method is not super long and it makes it a bit easier to read.